[LLVM][OpenMP] Add "nowait" clause as valid for "workshare"

kparzysz opened the pull request:
Add the "nowait" clause to the list of allowed clauses for the "workshare"
directive. This will make it consistent with other directives (which are
shared between C/C++ and Fortran).

The parser will still reject "nowait" on "!$omp workshare", so this has no
effect on accepting/rejecting Fortran source code.

kparzysz committed:
[LLVM][OpenMP] "nowait" clause is valid for "workshare"

Add the "if" clause to the list of allowed clauses for the "workshare"
directive.

carlobertolli reviewed (approved):
LGTM: nowait has been part of OpenMP since 4.5 (or earlier, I didn't check 4.0).
It'd be nice if you added a clang/test/OpenMP minimal test. I checked and I couldn't find a single parse test for the workshare directive.

kparzysz commented:
I can add a flang test for it (clang won't likely handle it, since "workshare" is Fortran-only).

kiranchandramohan reviewed (approved):
LG.

kiranchandramohan reviewed (changes requested):
Sorry. It is already there in "End Workshare".

kparzysz commented:
> Sorry. It is already there in "End Workshare".

All directives that have "DIR/END DIR" format in Fortran, allow the clauses from the "END DIR" to be present in "DIR" (since there is no "END DIR" in C/C++).  Workshare is be the only one that doesn't allow it.

Not having the consistency makes it harder to handle OpenMP constructs with language-agnostic code.

kiranchandramohan reviewed (approved):
That makes sense. Please wait for @mjklemm or @kkwli before proceeding. If you are going ahead might be good to call this out in https://github.com/llvm/llvm-project/blob/main/flang/docs/Extensions.md

Looking at other compilers. gfortran seems to accept it, ifx and classic flang rejects.

> Not having the consistency makes it harder to handle OpenMP constructs with language-agnostic code.

It wasn't clear from the title that this is deviating from the standard specification for ease of implementation.

> Add the "if" clause to the list of allowed clauses for the "workshare" directive.

"if" -> "nowait"

kkwli commented:
As far as I understand, `nowait` is only allowed on the `end workshare` directive in the spec. I have no objection to extend it to have the clause on the `workshare` directive as long as it also allowed on the `end workshare` directive. However, we probably want to add semantic check for having `nowait` on both `workshare` and `end workshare`. How about the `copyprivate` clause on the `single` directive?

kparzysz commented:
> It wasn't clear from the title that this is deviating from the standard specification for ease of implementation.

It was a good catch on your side.  I was testing my code that relied on this (which I had forgotten about).  When I created the PR I honestly thought that this is where the "nowait" clause was supposed to be.  Only after you pointed out the "end workshare" when I realized what really happened.

kparzysz committed:
Disallow "nowait" on "!$omp workshare" in parser

kparzysz commented:
> However, we probably want to add semantic check for having `nowait` on both `workshare` and `end workshare`. How about the `copyprivate` clause on the `single` directive?

The flang parser actually checks for such cases explicitly, and rejects cases of "nowait" on directives that allow it based on the .td contents, but where Fortran syntax requires them to be on the "end" directive.  I added "workshare" to the list, and updated the testcase.  I have checked that "copyprivate" on "!$omp single" is rejected as well.

In the current form this PR will not have any effect on the accepted syntax, it will simply allow "nowait" to be associated with "workshare" internally to the compiler.

mjklemm commented:
> As far as I understand, `nowait` is only allowed on the `end workshare` directive in the spec.

Post-merge comment: this was changed with OpenMP 5.2:

![image](https://github.com/llvm/llvm-project/assets/5994372/157be687-b823-49b7-b53f-ab96c7b1b296)
